### PR TITLE
[Resolves #894] Switch to using poetry

### LINKED ISSUE #894: Implement package management and accommodate PEP-517/518
# Package Management / Modernization

## Context

Currently, sceptre is using the setup.py, setup.cfg, and dev/prod requirements.txt files to handle sceptre packaging, dependencies, and publishing.

[PEP-517](https://www.python.org/dev/peps/pep-0517/)
[PEP-518](https://www.python.org/dev/peps/pep-0518/)
[PEP 517 and 518 in Plain English](https://medium.com/@grassfedcode/pep-517-and-518-in-plain-english-47208ca8b7a6)

## Possible Fix

Migrate to use [Poetry](https://github.com/python-poetry/poetry), removing setup.py and MANIFEST.in in favor of the new pyproject.toml standard.

### PR DESCRIPTION
**Pull request by zaro0508**
The intent here is to change the tooling over to use [poetry](https://python-poetry.org) instead of `setuptools`.
This is building on the work that was started by @tarkatronic in PR #1035

Some important things to note:

* Poetry will be used for packaging, dependencies, and publishing. 
* Poetry is smart enough to automatically include data files such as the `*.json` that had to be manually included by setuptools.
* `twine` is no longer necessary as we can simply use `poetry publish`.
* Poetry handles all virtualenv matters automatically. So we no longer need to instruct users to construct virtualenvs for development purposes, nor do we need to manually create and manage them for CI.
* Dependency specifications are a bit different with Poetry, and a bit more "modern", [as seen here](https://python-poetry.org/docs/dependency-specification/).
* I have told git (via the `.gitattributes`) file to treat the `poetry.lock` file as binary. This is a fairly standard practice, also followed in the JS community with the `package-lock.json` file. This does two primary things:
  * Tells GitHub to not attempt to display a diff for this file
  * Prevents frequent merge conflicts, since this tends to be a frequently updated file

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1 @@
+poetry.lock binary
```
Comment (jfalkenstein) [inline .gitattributes:1]: What exactly does this do?

File: pyproject.toml [hunk 2]
```diff
@@ -1,30 +1,88 @@
...
+"split" = "sceptre.resolvers.split:Split"
+"join" = "sceptre.resolvers.join:Join"
+
+[tool.poetry.plugins."sceptre.template_handlers"]
+"file" = "sceptre.template_handlers.file:File"
+"s3" = "sceptre.template_handlers.s3:S3"
+"http" = "sceptre.template_handlers.http:Http"
+
+[tool.poetry.dependencies]
+python = ">=3.8,<3.12"
```
Comment (jfalkenstein) [inline pyproject.toml:49]: This officially drops support for 3.7. I don't think we should do that yet. Let's wait until it hits EOL.

File: pyproject.toml [hunk 3]
```diff
@@ -1,30 +1,88 @@
...
+boto3 = ">=1.3.0,<2"
+click = ">=7.0,<9.0"
+cfn-flip = ">=1.2.3,<2.0"
```
Comment (jfalkenstein) [inline pyproject.toml:52]: We can use the shortened syntax for this
```suggestion
cfn-flip = "^1.2.3"
```

File: pyproject.toml [hunk 4]
```diff
@@ -1,30 +1,88 @@
...
+deepdiff = ">=5.5.0,<6.0"
+deprecation = ">=2.0.0,<3.0"
+jinja2 = ">=3.0,<4"
```
Comment (jfalkenstein) [inline pyproject.toml:55]:
```suggestion
jinja2 = "^3"
```

File: pyproject.toml [hunk 5]
```diff
@@ -1,30 +1,88 @@
...
+jsonschema = ">=3.2,<3.3"
+networkx = ">=2.6,<2.7"
+packaging = ">=16.8,<22.0"
+pyyaml = ">=5.1,<6.0"
```
Comment (jfalkenstein) [inline pyproject.toml:59]:
```suggestion
pyyaml = "^5.1"
```

File: pyproject.toml [hunk 6]
```diff
@@ -1,30 +1,88 @@
...
+sceptre-cmd-resolver = ">=1.1.3,<3"
+sceptre-file-resolver = ">=1.0.4,<2"
```
Comment (jfalkenstein) [inline pyproject.toml:61]:
```suggestion
sceptre-file-resolver = "^1.0.4"
```

File: pyproject.toml [hunk 7]
```diff
@@ -1,30 +1,88 @@
...
+colorama = ">=0.2.5,<0.4.4"
+troposphere = { version = ">4,<5", optional = true }
+
+[tool.poetry.extras]
+runtime = ["troposphere"]
```
Comment (jfalkenstein) [inline pyproject.toml:66]: So what's going on here? How does this combine with the `troposphere` above?

File: pyproject.toml [hunk 8]
```diff
@@ -1,30 +1,88 @@
...
+
+[tool.poetry.group.dev.dependencies]
+pre-commit = "^2.21.0"
+behave = ">=1.2.5,<2.0.0"
```
Comment (jfalkenstein) [inline pyproject.toml:70]:
```suggestion
behave = "^1.2.5"
```

File: pyproject.toml [hunk 9]
```diff
@@ -1,30 +1,88 @@
...
+freezegun = ">=0.3.12,<0.4.0"
+moto = ">=3.0,<4"
```
Comment (jfalkenstein) [inline pyproject.toml:72]:
```suggestion
moto = "^3"
```

File: pyproject.toml [hunk 10]
```diff
@@ -1,30 +1,88 @@
...
+pygments = ">=2.2.0,<3.0.0"
```
Comment (jfalkenstein) [inline pyproject.toml:73]:
```suggestion
pygments = "^2.2"
```

File: pyproject.toml [hunk 11]
```diff
@@ -1,30 +1,88 @@
...
+pytest = ">=6.2.0,<7.0.0"
```
Comment (jfalkenstein) [inline pyproject.toml:74]:
```suggestion
pytest = "^6.2"
```

File: pyproject.toml [hunk 12]
```diff
@@ -1,30 +1,88 @@
...
+pytest-cov = ">=2.11.1,<3.0.0"
+pytest-sugar = ">=0.9.4,<1.0.0"
```
Comment (jfalkenstein) [inline pyproject.toml:76]:
```suggestion
pytest-sugar = "^0.9.4"
```

File: pyproject.toml [hunk 13]
```diff
@@ -1,30 +1,88 @@
...
+readme-renderer = ">=24.0,<25.0"
```
Comment (jfalkenstein) [inline pyproject.toml:77]:
```suggestion
readme-renderer = "^24"
```

File: pyproject.toml [hunk 14]
```diff
@@ -1,30 +1,88 @@
...
+requests-mock = ">=1.9.3,<2.0"
```
Comment (jfalkenstein) [inline pyproject.toml:78]:
```suggestion
requests-mock = "^1.9.3"
```

File: pyproject.toml [hunk 15]
```diff
@@ -1,30 +1,88 @@
...
+sphinx = ">=1.6.5,<=5.1.1"
+sphinx-click = ">=2.0.1,<4.0.0"
+sphinx-rtd-theme = "0.5.2"
+sphinx-autodoc-typehints = "1.19.2"
+docutils = "<0.17"
+tox = ">=3.23.0,<4.0.0"
```
Comment (jfalkenstein) [inline pyproject.toml:84]:
```suggestion
tox = "^3.23.0"
```

File: sceptre/__init__.py [hunk 16]
```diff
@@ -3,11 +3,12 @@
...
 import warnings
+import importlib.metadata
 
 
-__author__ = "Cloudreach"
-__email__ = "sceptre@cloudreach.com"
-__version__ = "4.0.2"
+__author__ = "SceptreOrg"
+__email__ = "sceptreorg@gmail.com"
+__version__ = importlib.metadata.version(__package__ or __name__)
```
Comment (jfalkenstein) [inline sceptre/__init__.py:11]: Interesting!

File: pyproject.toml [hunk 17]
```diff
@@ -1,30 +1,88 @@
...
-exclude_lines = [
-  # Have to re-enable the standard pragma
-  "pragma: no cover", # Don't complain about missing debug-only code:
-  "def __repr__",
-  "if self.debug", # Don't complain if tests don't hit defensive assertion code:
-  "raise AssertionError",
-  "raise NotImplementedError", # Don't complain if non-runnable code isn't run:
-  "if 0:",
-  "if __name__ == .__main__.:", # Don't complain about mypy-specific code
-  "if TYPE_CHECKING:",
```
Comment (jfalkenstein) [inline pyproject.toml:19]: Why are you removing all these other tool configurations?

File: pyproject.toml [hunk 18]
```diff
@@ -1,30 +1,88 @@
...
-ignore_errors = true
-fail_under = 90.0
-show_missing = true
-
-[tool.pytest.ini_options]
-# The "--cov" option will disable the pycharm debugger.  To enable the debugger either
-# remove this flag or add the "--no-cov" to your pycharm's pytest run configuration.
-# https://github.com/pytest-dev/pytest-cov/issues/131
-# https://stackoverflow.com/questions/34870962/how-to-debug-py-test-in-pycharm-when-coverage-is-enabled
-addopts = "-s --cov --cov-report term --cov-report term-missing --cov-report xml"
```
Comment (jfalkenstein) [inline pyproject.toml:30]: Why are you removing this?

File: Makefile [hunk 19]
```diff
@@ -1,116 +0,0 @@
-.PHONY: clean-pyc clean-build docs clean docs
```
Comment (jfalkenstein) [inline Makefile:1]: Why are you removing this? This is pretty handy to have...

File: .circleci/config.yml [hunk 20]
```diff
@@ -38,70 +38,44 @@ jobs:
...
-            - sceptre-{{ .Environment.CACHE_VERSION }}-dependencies-{{ arch }}-{{
-              checksum "requirements/prod.txt" }}-{{ checksum
-              "requirements/dev.txt" }}-{{ checksum "tox.ini" }}
+            - sceptre-v1-{{ checksum "pyproject.toml" }}-{{ checksum "poetry.lock" }}
+            - sceptre-v1-{{ checksum "pyproject.toml" }}-{{ checksum "poetry.lock" }}-{{ checksum "tox.ini" }}
       - run:
-          name: 'Linting'
-          command: |
-            pyenv global venv
-            make lint
```
[the remark this answers is not in the document]
Comment (zaro0508) [reply .circleci/config.yml:90]: this was just calling `pre-commit` to do the linting.  I replaced it with a direct call to `poetry run pre-commit run --all-files --show-diff-on-failure` which does the same thing but within a poetry virtual env.

File: .gitattributes (hunk 1 above)
Comment (zaro0508) [reply .gitattributes:1]: the description is in the PR description
```
I have told git (via the .gitattributes) file to treat the poetry.lock file as binary. This is a fairly standard practice, also followed in the JS community with the package-lock.json file. This does two primary things:
Tells GitHub to not attempt to display a diff for this file
Prevents frequent merge conflicts, since this tends to be a frequently updated file
```

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -181,27 +174,27 @@ during development.
...
+_Note_: All integration tests are set up to run in `eu-west-*` region.  If you prefer
 to run in a different region you must update the region in each test before running it
 
-### run all tests
+
+#### run all tests
 
 ```bash
-$ behave integration-tests
+$ AWS_PROFILE=<profile> AWS_DEFAULT_REGION=<region> poetry run behave --junit --junit-directory build/behave
```
[the remark this answers is not in the document]
Comment (zaro0508) [reply CONTRIBUTING.md:185]: no, it works fine in other regions but tests do need to be updated a little to make it work.  There is a little blurb on it in the docs.. `If you prefer to run in a different region you must update the region in each test before running it`

File: Dockerfile [hunk 22]
```diff
@@ -1,7 +1,7 @@
-FROM python:3.7-alpine
+FROM python:3.8-alpine
```
[the remark this answers is not in the document]
Comment (zaro0508) [reply Dockerfile:1]: hmm, sure that's probably a good idea.

File: Makefile (hunk 19 above)
Comment (zaro0508) [reply Makefile:1]: everything should be done using the poetry CLI now so the purpose here is to eliminate a tool that's not needed anymore.

File: pyproject.toml (hunk 3 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml [hunk 23]
```diff
@@ -1,30 +1,88 @@
...
+"file" = "sceptre.template_handlers.file:File"
+"s3" = "sceptre.template_handlers.s3:S3"
+"http" = "sceptre.template_handlers.http:Http"
+
+[tool.poetry.dependencies]
+python = ">=3.8,<3.12"
+boto3 = ">=1.3.0,<2"
+click = ">=7.0,<9.0"
+cfn-flip = ">=1.2.3,<2.0"
+deepdiff = ">=5.5.0,<6.0"
```
[the remark this answers is not in the document]
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml [hunk 24]
```diff
@@ -1,30 +1,88 @@
...
+deprecation = ">=2.0.0,<3.0"
```
[the remark this answers is not in the document]
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml (hunk 4 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml (hunk 7 above)
Comment (zaro0508) [reply pyproject.toml:66]: done

File: pyproject.toml (hunk 8 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml (hunk 9 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml (hunk 14 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml (hunk 6 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml (hunk 10 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml (hunk 13 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml (hunk 15 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: pyproject.toml (hunk 2 above)
Comment (jfalkenstein) [reply pyproject.toml:49]: If we drop support for 3.7, I'm wondering if the next version ought to be a major version...

**Comment by zaro0508** [comment]
@jfalkenstein i got unit tests working for python 3.7 now as well.

**Comment by jfalkenstein** [comment]
I hope to test this some evening this week. Sorry for the delay.

File: pyproject.toml [hunk 25]
```diff
@@ -1,30 +1,88 @@
...
+deprecation = "^2.0"
+jinja2 = "^3.0"
+jsonschema = ">=3.2,<3.3"
+networkx = ">=2.6,<2.7"
+packaging = ">=16.8,<22.0"
+pyyaml = "^5.1"
+sceptre-cmd-resolver = "^2.0"
+sceptre-file-resolver = "^1.0"
+colorama = ">=0.2.5,<0.4.4"
+troposphere = { version = ">4,<5", optional = true }
```
Comment (jfalkenstein) [inline pyproject.toml:63]: I think you should add troposphere as a dev-dependency, since unit tests can't fully run without it.
```suggestion
troposphere = { version = "^4", optional = true }
```

File: pyproject.toml (hunk 7 above)
Comment (jfalkenstein) [reply pyproject.toml:66]: Why is this extra called "runtime"? This means to install this extra explicitly you need to invoke `poetry install -E runtime` which doesn't make sense. I think it should be called `troposphere`.

**Review by jfalkenstein — CHANGES_REQUESTED** [review]
I think the troposphere extra is implemented improperly (or at least weirdly). It should also be a dev dependency. 

Also, the github pull request template references running `make test`. 

Finally, I think it's a pretty significant loss with losing the makefile. I really feel like the makefile didn't have to be removed and it should have just been updated.

File: Makefile (hunk 19 above)
Comment (jfalkenstein) [reply Makefile:1]: I don't think that the need for this file has gone away. Some of this file can go away, sure. But this file gave a very simple interface for a lot of common functionality. We can still have this file invoke `poetry run ....` but I'd rather keep the makefile.

File: .pre-commit-config.yaml [hunk 26]
```diff
@@ -44,6 +44,12 @@ repos:
         # pre-commit's default_language_version, see
         # https://pre-commit.com/#top_level-default_language_version
         language_version: python3.10
+  - repo: https://github.com/python-poetry/poetry
+    rev: '1.4.0'
+    hooks:
+      - id: poetry-check
+      - id: poetry-lock
+        language_version: python3.10
```
Comment (jfalkenstein) [inline .pre-commit-config.yaml:52]: Add this if you don't want a frustrating experience with this pre-commit hook
```suggestion
        language_version: python3.10
        args: ['--check']
```

**Comment by jfalkenstein** [comment]
It also appears as if integration tests are failing on this.

File: Makefile (hunk 19 above)
Comment (zaro0508) [reply Makefile:1]: lets replace the makefile with [poetry aliases](https://pypi.org/project/poetry-aliases). this PR is already pretty large so lets do it as a follow on PR?

File: pyproject.toml (hunk 25 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: .pre-commit-config.yaml (hunk 26 above)
Comment (zaro0508) [reply .pre-commit-config.yaml:52]: done

File: pyproject.toml (hunk 7 above)
Comment (zaro0508) [reply pyproject.toml]: done

File: Makefile (hunk 19 above)
Comment (jfalkenstein) [reply Makefile:1]: Poetry aliases are not project specific. It appears they are more like git aliases and set up globally. They won't work for this use case.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 27]
```diff
@@ -6,9 +6,9 @@
 - [ ] Commit message starts with `[Resolve #issue-number]`.
 - [ ] Added/Updated unit tests.
 - [ ] Added/Updated integration tests (if applicable).
-- [ ] All unit tests (`make test`) are passing.
+- [ ] All unit tests (`poetry run tox`) are passing.
 - [ ] Used the same coding conventions as the rest of the project.
-- [ ] The new code passes pre-commit validations (`pre-commit run --all-files`).
+- [ ] The new code passes pre-commit validations (`poetry pre-commit run --all-files`).
```
Comment (jfalkenstein) [inline .github/PULL_REQUEST_TEMPLATE.md:11]: I think `poetry run pre-commit run --all-files` is the correct command.

Comment (zaro0508) [reply .github/PULL_REQUEST_TEMPLATE.md]: fixed

File: pyproject.toml [hunk 28]
```diff
@@ -1,30 +1,97 @@
...
-  "def __repr__",
-  "if self.debug", # Don't complain if tests don't hit defensive assertion code:
-  "raise AssertionError",
-  "raise NotImplementedError", # Don't complain if non-runnable code isn't run:
-  "if 0:",
-  "if __name__ == .__main__.:", # Don't complain about mypy-specific code
-  "if TYPE_CHECKING:",
+[tool.poetry]
+name = "sceptre"
+version = "4.0.2"
```
Comment (zaro0508) [inline pyproject.toml:3]: need to update to 4.1.0

**Comment by jfalkenstein** [comment]
I'll try to test this out tonight.

File: Makefile [hunk 29]
```diff
@@ -48,24 +48,24 @@ clean-test:
...
 
 test-integration: install
-	behave integration-tests/
+	poetry run behave integration-tests/
 
 docs:
 	rm -f docs/sceptre.rst
 	rm -f docs/modules.rst
-	sphinx-apidoc -o docs/ sceptre
+	poetry run make docs
```
Comment (jfalkenstein) [inline Makefile:68]: I think this was a whoopsie. Running `make docs` infinitely calls `poetry run make docs`. 
```suggestion
	poetry run sphinx-apidoc -o docs/ sceptre
```

**Comment by jfalkenstein** [comment]
This looks good to me now. I'm running integration tests on it now. Assuming all tests pass, I'll approve.

**Review by jfalkenstein — APPROVED** [review]
Congratulations.
